system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,526 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+#include <cstdlib>
+#include <queue>
+#include <typeinfo>
+#include "mpi_wrapper.hpp"
+#include "mputil.hpp"
+
+#define WAIT_DELAY 100
+#define PROBING_DELAY 100
+
+//----------Functions and Data structures managing communication data related to Send/Recv Communications in orogress-----------//
+
+// Data structure to keep the data relating to send/receive communications
+class CommData
+{
+private:
+    bool send;                      // TRUE => relates to a send communication | FALSE => relates to receive communication
+    bool locked = false;
+    bool cancellationLock = false;
+    bool cancellationInProgress = false;
+    CriticalSection dataChangeLock;
+    void lock(){dataChangeLock.enter(); locked=true;}
+    void unlock(){locked=false; dataChangeLock.leave();}
+public:    
+    bool isSend(){return send;}
+    bool isReceive(){return !isSend();}
+    bool isEqual(bool _send, int _rank, int _tag)
+    {
+        return (send==_send) && (_rank==MPI_ANY_SOURCE || rank==_rank) && (_tag==MPI_ANY_TAG || tag==_tag);
+    }
+
+    int rank;                       // source/destination rank of the processor
+    int tag;                        // MPI tag information
+    MPI_Request* request;           // persistent request object to keep track of ongoing MPI call
```

Comment:
Yeah, we can use either. I put a pointer because whenever this is used in MPI functions it requires a pointer to the MPI_Request object. I was going to make this a private variable as you've suggested and introduce a function to return the memory pointer instead. but ended up forgetting to do that. I will get that done.